### PR TITLE
Feat/gas refund count safety module stakes

### PR DESCRIPTION
**Pull request by mwamedacen**
Implementing the resolution of safety module stakes by feeding an in memory timeseries data structure on listening key events.

Glossary:
- [StkPSPbpt](https://etherscan.io/address/0xc8dc2ec5f5e02be8b37a8444a1931f02374a17ab): Safety Module of ParaSwap (and an ERC20). It exchanges at 1:1 with underlying BPT token
- [BPT](http://etherscan.io/address/0xcb0e14e96f2cefa8550ad8e4aea344f211e5061d): 80% PSP - 20% WETH Balancer pool (and an ERC20)
- [BVault](https://etherscan.io/address/0xba12222222228d8ba445958a75a0704d566bf2c8): balancer vault, holds all funds and controls pools liquidity

To resolve the PSP stake of a single user we rely a simple formula: 
`StakedPSP(address) => PSP.balanceOf(BPT) / BPT.totalSupply() * StkPSPbpt.balanceOf(address)`

All members of this formula are timeseries here. 

Hints for reviewing:
- PSP.balanceOf(BPT) is given by Bvault.getTokensInfo() + PoolBalanceChanged/Swap events.
- Follow advice in the test [comments](https://github.com/paraswap/paraswap-volume-tracker/blob/49c5d7b0a289df82f8691cfe3624a5d69962efa9/__tests__/gas-refund-program/safety-module-stakes-tracker.ts) 
- Each epoch takes now ~15min to process. I advise commenting `Database.transaction()` to see data flowing in your tables

### CODE REVIEW

File: scripts/gas-refund-program/staking/safety-module-stakes-tracker.ts [hunk 1]
```diff
@@ -0,0 +1,449 @@
...
+} from '../timeseries';
+
+const SafetyModuleAddress = '0xc8dc2ec5f5e02be8b37a8444a1931f02374a17ab';
+const BalancerVaultAddress = '0xba12222222228d8ba445958a75a0704d566bf2c8';
+const Balancer_80PSP_20WETH_poolId =
+  '0xcb0e14e96f2cefa8550ad8e4aea344f211e5061d00020000000000000000011a';
+const Balancer_80PSP_20WETH_address = Balancer_80PSP_20WETH_poolId.substring(
+  0,
+  42,
+); // or 0xcb0e14e96f2cefa8550ad8e4aea344f211e5061d
```
Comment (alexshchur) [inline scripts/gas-refund-program/staking/safety-module-stakes-tracker.ts:28]:
```suggestion
const Balancer_80PSP_20WETH_address = '0xcb0e14e96f2cefa8550ad8e4aea344f211e5061d'
const Balancer_80PSP_20WETH_poolId = `${Balancer_80PSP_20WETH_address}00020000000000000000011a`
```

maybe this'll be a bit more intuitive. Or maybe not

File: scripts/gas-refund-program/persistance/db-persistance.ts [hunk 2]
```diff
@@ -85,6 +85,38 @@ export async function fetchTotalRefundedAmountUSDByAddress(): Promise<{
...
+    raw: true,
+  })) as unknown as { chainId: number; lastTimestampForChain: number }[];
+
+  const lastTxTimestampsAllChains = chainToTxTimestamp.map(
+    t => t.lastTimestampForChain,
+  );
+  const latestTxTimestamps = Math.min(...lastTxTimestampsAllChains);
+
+  return latestTxTimestamps;
+}
```
Comment (samthomson) [inline scripts/gas-refund-program/persistance/db-persistance.ts:118]: handle no previous `GasRefundParticipation` records, scenario

**Comment by samthomson** [comment]
left some comments. code seems more or less ok, but I haven't yet been able to run it through successfully.

**Review by samthomson — COMMENTED** [review]
left some comments. code seems more or less ok, but I haven't yet been able to run it through successfully.

File: __tests__/gas-refund-program/safety-module-stakes-tracker.ts [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    await tracker.loadStakes(false);
+  });
+
+  test('Init state', () => {
+    // hint: Use https://etherscan.io/tokencheck-tool to check all snapshoted data
+    const initStateSerialised = {
+      bptPoolPSPBalance: tracker.initState.bptPoolPSPBalance.toFixed(), // hint: given by balancer vault by looking at getPoolTokenInfo() - execute rpc call at specific block
+      bptPoolTotalSupply: tracker.initState.bptPoolTotalSupply.toFixed(), // hint: check total supply of 0xcb0e14e96f2cefa8550ad8e4aea344f211e5061d at block
+      stkPSPBptUsersBalances: Object.fromEntries(
+        // hint: check balances of 0xc8dc2ec5f5e02be8b37a8444a1931f02374a17ab of users at blocl
```
Comment (Velenir) [inline __tests__/gas-refund-program/safety-module-stakes-tracker.ts:25]:
```suggestion
        // hint: check balances of 0xc8dc2ec5f5e02be8b37a8444a1931f02374a17ab of users at block
```

File: scripts/gas-refund-program/transactions-indexing/swaps-subgraph.ts [hunk 4]
```diff
@@ -74,17 +70,23 @@ export async function getSwapsForAccounts({
...
+  let skip = 0;
+  let pageSize = 100;
+
+  while (true) {
+    const _swaps = await query(skip, pageSize);
+    swaps = swaps.concat(_swaps);
+    if (_swaps.length < pageSize) {
+      break;
+    }
+    skip = skip + pageSize;
```
Comment (Velenir) [inline scripts/gas-refund-program/transactions-indexing/swaps-subgraph.ts:86]: Does pagination always run ok? I seem to remember there were some problems with Graph pagination...

**Review by Velenir — COMMENTED** [review]
Code looks good, can't run right now to check though.

File: scripts/gas-refund-program/transactions-indexing/swaps-subgraph.ts (hunk 4 above)
Comment (mwamedacen) [reply scripts/gas-refund-program/transactions-indexing/swaps-subgraph.ts:86]: It's true that thegraph pagination is not reliable at some point. But this is only a secondary loop.
Screenshot 2022-05-02 at 12 48 03

The primary loop is sliding a 6h window [here](https://github.com/paraswap/paraswap-volume-tracker/blob/cf913fdfb8675b8ed7e31aa6a47a2ea59f819b75/scripts/gas-refund-program/transactions-indexing/successful-swap-tx-indexing.ts#L62). 

The secondary (this one), scans 100 transactions in batch.

The assumption is that we won't scan more than 5000 tx within 6h. If this goes wrong, first the script will crash. We can always lower the primary loop cycle.

I didn't see it crash with our conditions though + we're migrating to another source soon or later. I'm open to suggestion

File: scripts/gas-refund-program/computeGasRefund.ts [hunk 5]
```diff
@@ -31,21 +33,15 @@ async function startComputingGasRefundAllChains() {
     await GRPSystemGuardian.loadStateFromDB();
     GRPSystemGuardian.assertMaxPSPGlobalBudgetNotReached();
 
+    await SafetyModuleStakesTracker.getInstance().loadStakes();
```
[the remark this answers is not in the document]
Comment (mwamedacen) [reply scripts/gas-refund-program/computeGasRefund.ts:36]: getInstance should be the way as it eases the test. Will change guardian later.

File: scripts/gas-refund-program/staking/safety-module-stakes-tracker.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (mwamedacen) [reply scripts/gas-refund-program/staking/safety-module-stakes-tracker.ts:28]: your method is nice guys but this is error prone. I prefer a full formed poolId string to check and then composing string s and end up with wrong poolId due to typo whatever.

File: scripts/gas-refund-program/transactions-indexing/successful-swap-tx-indexing.ts [hunk 6]
```diff
@@ -105,190 +100,191 @@ export async function computeSuccessfulSwapsTxFeesRefund({
...
-        return;
-      }
+          swapperStake = sPSPStake.plus(safetyModuleStake);
+        } else {
+          swapperStake = sPSPStake;
+        }
 
-      const currencyRate = resolvePrice(+swap.timestamp);
+        if (swapperStake.isZero()) {
+          return;
```
[the remark this answers is not in the document]
Comment (mwamedacen) [reply scripts/gas-refund-program/transactions-indexing/successful-swap-tx-indexing.ts:157]: Not sure I got you, we need the total of PSP staked in sPSP + safetyModule to know if a tx is eligible for refund

File: scripts/gas-refund-program/transactions-indexing/swaps-subgraph.ts (hunk 4 above)
Comment (alexshchur) [reply scripts/gas-refund-program/transactions-indexing/swaps-subgraph.ts:86]: I can confirm `swaps` array ends up with some duplications in it. Not yet sure whether it's harmful or not, didn't yet follow the logic further

**Review by alexshchur — APPROVED** [review]
couldn't get to thinking of and testing edge cases, but in general code looks ok and quick tests I did on my side (and verifications of time series data Mohamed showed me how to do) worked fine. Also because for the nearest epoch it doesn't make big difference, I think we should deploy it so that a nice announcement can already be done to the community

**Review by samthomson — APPROVED** [review]
